src/Build/Logging/BinaryLogger/ProjectImportsCollector.cs

```diff
@@ -0,0 +1,140 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.IO.Compression;
+using System.Threading.Tasks;
+
+namespace Microsoft.Build.Logging
+{
+    /// <summary>
+    /// Creates a zip archive with all the .csproj and .targets encountered during the build.
+    /// The internal .zip file structure matches closely the layout of the original sources on disk.
+    /// The .zip file can be used to correlate the file names and positions in the build log file with the
+    /// actual sources.
+    /// </summary>
+    internal class ProjectImportsCollector
+    {
+        private FileStream _fileStream;
+        private ZipArchive _zipArchive;
+
+        public string ArchiveFilePath { get; set; }
+
+        /// <summary>
+        /// Avoid visiting each file more than once.
+        /// </summary>
+        private readonly HashSet<string> _processedFiles = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
+
+        // this will form a chain of file write tasks, running sequentially on a background thread
+        private Task _currentTask = Task.CompletedTask;
+
+        public ProjectImportsCollector(string logFilePath, string sourcesArchiveExtension = ".projectimports.zip")
+        {
+            ArchiveFilePath = Path.ChangeExtension(logFilePath, sourcesArchiveExtension);
+
+            try
+            {
+                _fileStream = new FileStream(ArchiveFilePath, FileMode.Create, FileAccess.ReadWrite, FileShare.Delete);
+                _zipArchive = new ZipArchive(_fileStream, ZipArchiveMode.Create);
+            }
+            catch
+            {
+                // For some reason we weren't able to create a file for the archive.
+                // Disable the file collector.
+                _fileStream = null;
+                _zipArchive = null;
+            }
+        }
+
+        public void AddFile(string filePath)
+        {
+            if (filePath == null || _fileStream == null)
+            {
+                return;
```

Comment:
I think it should be an error. It is not a valid program state. You can use `ErrorUtilities.VerifyThrowInternalNull`